ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionProcessor.java

```diff
@@ -295,77 +296,61 @@ public TransactionProcessingResult processTransaction(
-                .initialGas(gasAvailable)
-                .address(to)
-                .originator(senderAddress)
-                .contract(to)
-                .contractAccountVersion(
-                    contract != null ? contract.getVersion() : Account.DEFAULT_VERSION)
-                .gasPrice(transactionGasPrice)
-                .inputData(transaction.getPayload())
-                .sender(senderAddress)
-                .value(transaction.getValue())
-                .apparentValue(transaction.getValue())
-                .code(new Code(contract != null ? contract.getCode() : Bytes.EMPTY))
-                .blockHeader(blockHeader)
-                .depth(0)
-                .completer(c -> {})
-                .miningBeneficiary(miningBeneficiary)
-                .blockHashLookup(blockHashLookup)
-                .maxStackSize(maxStackSize)
-                .isPersistingPrivateState(isPersistingPrivateState)
-                .transactionHash(transaction.getHash())
-                .privateMetadataUpdater(privateMetadataUpdater)
-                .build();
-      }
+      final MessageFrame.Builder commonMessageFrameBuilder =
+          MessageFrame.builder()
+              .messageFrameStack(messageFrameStack)
+              .maxStackSize(maxStackSize)
+              .returnStack(new ReturnStack())
+              .blockchain(blockchain)
+              .worldState(worldUpdater.updater())
+              .initialGas(gasAvailable)
+              .originator(senderAddress)
+              .gasPrice(transactionGasPrice)
+              .sender(senderAddress)
+              .value(transaction.getValue())
+              .apparentValue(transaction.getValue())
+              .blockHeader(blockHeader)
+              .depth(0)
+              .completer(__ -> {})
+              .miningBeneficiary(miningBeneficiary)
+              .blockHashLookup(blockHashLookup)
+              .isPersistingPrivateState(isPersistingPrivateState)
+              .transactionHash(transaction.getHash())
+              .privateMetadataUpdater(privateMetadataUpdater);
+
+      final MessageFrame initialFrame =
+          transaction
+              .getTo()
+              .map(
+                  to -> {
+                    final Optional<Account> maybeContract = Optional.ofNullable(worldState.get(to));
+                    return commonMessageFrameBuilder
+                        .type(MessageFrame.Type.MESSAGE_CALL)
+                        .address(to)
+                        .contract(to)
+                        .contractAccountVersion(
+                            maybeContract
+                                .map(AccountState::getVersion)
+                                .orElse(Account.DEFAULT_VERSION))
+                        .inputData(transaction.getPayload())
+                        .code(
+                            new Code(maybeContract.map(AccountState::getCode).orElse(Bytes.EMPTY)));
+                  })
+              .orElseGet(
```

Comment:
This reduces the readability.  The if block makes it clear that there is one path for contract creation and one for normal transactions.  Without that you need to have deep knowledge of the transactions that a contract creation TX lacks a to field.  I don't mind the common field refactoring but I think the over-use of optional here reduces the accessibility to future maintainers.

Reply:
Yeah, it's true that the `isContractCreation` method does help with understanding what's going on. I can change the uncommon part back to the if/else.